waspc/src/Analyzer/TypeChecker/Internal.hs

```diff
@@ -0,0 +1,232 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TupleSections #-}
+
+-- | This module implements the type rules defined by the wasplang document
+-- in two phases.
+--
+-- = Hoisting Declarations
+--
+-- First, the variables bound by each declaration statement, e.g.
+--
+-- @
+-- app Todo {
+--   title: "Todo App"
+-- }
+-- @
+--
+-- are created, without checking any types. In the above example, the bindings
+-- created would be @[("Todo", DeclType "app")]@.
+--
+-- = Statement Type Checking
+--
+-- In this second phase, the types of the argument to each declaration are checked
+-- to ensure they are valid for the declaration type. The implementation of the
+-- type inference rules is in "checkExpr", "unifyTypes", and "weaken".
+module Analyzer.TypeChecker.Internal
+  ( check,
+    hoistDeclarations,
+    checkAST,
+    checkStmt,
+    checkExpr,
+    unify,
+    unifyTypes,
+    weaken,
+  )
+where
+
+import Analyzer.Parser (AST)
+import qualified Analyzer.Parser as P
+import Analyzer.Type
+import Analyzer.TypeChecker.AST
+import Analyzer.TypeChecker.Monad
+import Analyzer.TypeChecker.TypeError
+import qualified Analyzer.TypeDefinitions as TD
+import Control.Arrow (left)
+import Control.Monad (foldM)
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as M
+import Data.List.NonEmpty (NonEmpty ((:|)), nonEmpty, toList)
+
+check :: AST -> T TypedAST
+check ast = hoistDeclarations ast >> checkAST ast
+
+hoistDeclarations :: AST -> T ()
+hoistDeclarations (P.AST stmts) = mapM_ hoistDeclaration stmts
+  where
+    hoistDeclaration :: P.Stmt -> T ()
+    hoistDeclaration (P.Decl typName ident _) = setType ident $ DeclType typName
+
+checkAST :: AST -> T TypedAST
+checkAST (P.AST stmts) = TypedAST <$> mapM checkStmt stmts
+
+checkStmt :: P.Stmt -> T TypedStmt
+checkStmt (P.Decl typName name expr) =
+  lookupDeclType typName >>= \case
+    Nothing -> throw $ NoDeclarationType typName
+    Just (TD.DeclType _ expectedType) -> do
+      -- Decides whether the argument to the declaration has the correct type
+      mTypedExpr <- weaken expectedType <$> checkExpr expr
+      case mTypedExpr of
+        Left e -> throw e
+        Right typedExpr -> return $ Decl name typedExpr (DeclType typName)
+
+-- | Determine the type of an expression, following the inference rules described in
+-- the wasplang document. Some these rules are referenced by name in the comments
+-- of the following functions using [Brackets].
+checkExpr :: P.Expr -> T TypedExpr
+checkExpr (P.StringLiteral s) = return $ StringLiteral s
+checkExpr (P.IntegerLiteral i) = return $ IntegerLiteral i
+checkExpr (P.DoubleLiteral d) = return $ DoubleLiteral d
+checkExpr (P.BoolLiteral b) = return $ BoolLiteral b
+checkExpr (P.ExtImport n s) = return $ ExtImport n s
+checkExpr (P.Identifier ident) =
```

Comment:
I am still somewhat confused: I will try to repeat part of what you said that I understand and then explain my initial thoughts in the light of that.

I get that when creating tokens, a lot of tokens end up being lexed as identifiers, and those are all from the above that you described (app, Example, title, homePage, ...).

However, once we parse those into `Parser.AST`, they are not just identifiers any more, they become Dict, List, ... . Those that remain are references, which we call `Identifier` in `Parser.AST`, and that is `Home(1)`.
Then, as we transform `Parser.AST` to `TypeChecker.AST`, I wouldn't expect AST to change significantly -> references are still references and there are no other "identifiers", right? But, in `TypeChecker.AST` you have `Var` now instead of `Identifier`.

I agree that `Identifier` might not be the best name, but I feel `Var` feels too much like declaration. Also, we don't really have variables -> we have bindings. We don't have mutable code.

So what might be the best is to rename this `Identifier` to `Reference` and use that in both Parser.AST and TypeChecker.AST.

But, first we should try to clear up the situation with the many identifiers vs just Home(1) that you mentioned -> am I getting it right or not? Reduction from everything being identifier to just Home(1) being identifier is happening during transformation from token to parser AST, and not from parser AST to typechecked AST, correct?

Reply:
Yes, confusing!
- design doc -> sure, we can rename that if needed.
- item being called `Var` -> well that is an important argument. However, I would still not do it just because others do it, although I would based on that think twice about it. But for me, when choosing `Reference` and `Var`, `Reference` feels semantically more correct. So I would still prefer that, unless there is some other argument in favor of `Var`. What about you, which one do you like better and why?

Reply:
I still prefer `Var`. I think Haskell's naming is a good reference, since the semantics of defining/using variables in Wasp are (I believe) identical to Haskell's.

Reply:
I agree that Haskell is a good language to look at for inspiration, but I would be careful with assuming they made all the correct choices in naming, and would prefer double questioning why they did so - maybe we can do better.

Variable as a name is unusual to me for smth that is immutable - 'variable' means it can change! So why did Haskell use 'Var'? I don't know.

I googled a bit and read that in maths they also use term 'variable' even when it is single assignment (like in Haskell or Wasp). That helps a bit, it means people just ignore the fact that name means it is mutable. Still, it sounds like that is happening fmdue to lack of better naming.

Anyway, since it seems like Var is commonly used let's go with it, but I would feel better if we also had some reasoning to support why it is better than Reference.